toothpick-runtime/src/main/java/toothpick/UnScopedProviderImpl.java

```diff
@@ -44,60 +46,65 @@ public UnScopedProviderImpl(Class<?> factoryKeyClass, boolean isProviderFactoryC
 
   //we lock on the unbound provider itself to prevent concurrent usage
   //of the unbound provider (
-  public synchronized T get(Scope scope) {
-    if (instance != null) {
-      return instance;
-    }
-
-    if (providerInstance != null) {
-      if (isLazy) {
-        instance = providerInstance.get();
-        //gc
-        providerInstance = null;
+  public T get(Scope scope) {
+    lockGet.lock();
```

Comment:
there is bug is the article, so the test result may not true.
```
While I was reading your code I’ve realized the way you were using the ReentrantLock was wrong, because you were creating a new lock every time. If you are not using the same lock for each thread then it is similar to not having a lock. There is another comment pointing out this, but the bug is still there. It would be better if you could fix it, otherwise people might get confused if they do not read the comments.
```